### ISSUE TITLE
how to decrypt the data by a third part tool via SAS Enterprise Guide and SASPY

### ISSUE THREAD

**Issue by pradkum001** [sassoftware/saspy#405]
Hello All.,

I have been dealing with this issue directly with SAS. They have redirected me here. 

We have encrypted data in the database. The encryption is done by a third party tool. Now., the user wants to decrypt the data while they run a report or a query while using SAS enterprise guide.  

One of the teams in the organization have implemented the similar approach but using TabPy ( Tableau). They have created a TabPy script and call it in Tableau. While calling they pass the variables required to decrypt. This approach is working for them.

So, we wanted to use the same approach which for SAS is SASPy.  I have installed and configured SASPy. 

All i need to do in SAS enterprise guide is connect to the database and call the SASPy script and pass the variables which are needed to decrypt. 

Can someone help me on how to decrypt the data by a third part tool via SAS Enterprise Guide and SASPY. 

Thank you in advance.

**Comment by pradkum001** [comment]
HI Tom.,
Thank you for the quick response. After your reply, I just had a word with the tableau team to understand what exactly they are doing. I would like to mimic the exact same method what they have followed.

What they have done is they have built a TabPy function which is deployed on Tableau(using client.deploy command) and this function calls the REST API (Built on Java platform) which does all the *cryption. This process does not require any storage of data. Which means., after getting the desired results in tableau, they call this function along with the variables needed to decrypt. And that function is implemented and the values are decrypted and printed. 

In the same way., I would like to., First, create a SASPy function(similar to TABPy) which is deployed on sas server(using ?? commad) and run it from SAS Enterprise guide which calls the same REST API and the en|de cryption is triggered. 
In other words., When i run a query in SAS EG., the results are displayed., and out of those results assume that one variables needs to be decrypted. Now the SASPy functionality comes into picture which calls the REST API and decryption occurs. 

Please let me know if you have got the complete picture now? 

How can we deploy this SASPy function on the sas server (does sas has similar function as client.deploy please check this link: https://tableau.github.io/TabPy/docs/tabpy-tools.html)? 

And secondly., how can we call the saspy script via SAS EG.? (X "" command??) 

Also., please let me know if there is a better approach than this using the similar concept.

Thank you so much in advance.. Really Appreciate your help.

**Comment by tomweber-sas** [comment]
So, this *cryption software you want to use has a REST API? Then you can call it from either SAS or from Python. SAS has a Proc HTTP and Python has an http.client, or you can use Requests. I'm assuming you're reading and writing the data from the database using an access engine? If you use saspy, you can just get the values via symget()/symput() from/to SAS, since you you are saying they are just individual values. You can call the rest API from Python to encrypt/decrypt the values.
I'm not clear on whatever else your application does, or what all it's written in. SASPy is a Python module, so using it from Python  to interact with SAS is what it's for. It's not for creating functions to 'deploy' in SAS. Whatever tabpy happens to be or do, it's not the same thing as saspy and SAS.
If you're trying to run from EG, not python, it makes me think the rest of your application is in SAS, so you could just use proc htttp in SAS to call the *cryption functions, and it would all just be in your SAS program. Trying to use EG to call Python to use SASPy to call SAS seems convoluted. 
Are there any developers at your site who are familiar with SAS, and hopefully Python, who could help you with how to make these REST calls? It sounds like you just need to make them from your existing SAS code, really.

**Comment by pradkum001** [comment]
@Tomweber,

I agree and really like PROC HTTP is straightforward and easy., but for now, i would like to re-use the existing python script and call it from SAS EG because 1) The existing functionality is working and 2) it is an urgent request which needs to be deployed immediately. After deploying this concept, i can enhance it with PROC HTTP in the future.

So the approach is calling SASPy script from SAS EG., example:

I run a below query in SAS EG..

DATA temp;
SET db.patient(KEEP=patient_name patient_address Patient_ID);
RUN;

Where, patient_name and patient_address are the encrypted values and I need to decrypt those two variables from TEMP dataset by calling the script.

I am close to developing the SASPy script - currently working on installing some missing packages. Once its ready I can start integrating both. 

So my question now is (based on the above example)., 
1) how do we integrate SASPy script with SAS EG? 
2) And how do we call the the script from EG and pass those variables?
3) Are there any examples you can point me to..
4) Where do I save the SASPy script in order to execute it ? Can I create some kind of a macro and call it in a data step by passing a values?

Thanks alot Tom.

**Comment by tomweber-sas** [comment]
I'm not sure what the existing python script is. Is that what the tableau people wrote? The only relevant thing in that would be the python REST calls to en/de-crypt. That should be some simple python to do that.

I don't think you understand what saspy is yet. It is a Python client that connect to SAS (a remote workspace server in your environment). EG is a windows client that connects to a remote workspace server. EG is not SAS. EG is a client that submits SAS code to the SAS server it's connected to. SASPy is a client that submits SAS code to the SAS server it's connected to. They both get log and results and bring them back to the client . It doesn't make sense about trying to interface to saspy from EG.

You use saspy, from python, to connect to SAS and run SAS code, and you can interact between python (like to call your *crypt routines from python using SAS data values), by copying variables between SAS and Python, or moving data sets between SAS and Python. It's for mixed workflows that you want to write in Python. That sounds just like what it seems you want to do.

In your example, with the data step to get the columns to decrypt, you wouldn't submit that from EG, you would submit it from saspy, and then the data set you created can be pulled over to python and you can call the python decrypt method on those values. If you only have one row, you can just put the values in macro variables (in your data step) and then pull them over to python variables and decrypt then from python. you can send back encrypted values the same way, just in reverse; python variables to SAS macro variables or dataFrame to SAS Data Set. Your SAS code is run from saspy, same as your python code. EG has nothing to do with it.

The example notebook I pointed you at here (https://github.com/sassoftware/saspy-examples/blob/main/SAS_contrib/saspy_example_github.ipynb) has examples of submitting code and of moving data back and forth. The example notebook here (https://github.com/sassoftware/saspy-examples/blob/main/SAS_contrib/Using_SYMGET_and_SYMPUT.ipynb) has examples of moving variables between Python and SAS (macro variables).  The API doc has all of the methods documented showing what the signatures are and what each parameter is: https://sassoftware.github.io/saspy/api.html

I hope this is starting to make sense. Whatever that tabpy thing is and does from python, has nothing to do with how you use saspy to write a mixed Python/SAS workflow in Python.

Thanks,
Tom

**Comment by tomweber-sas** [comment]
FWIW, and this won't help you today, I think that in Viya 4 there is something that would work the way I think you're thinking about this. It's sounding to me like you want a SAS code mixed workflow w/ Python integration. So that from SAS, you can call some Python code, but from w/in your SAS code, where you could integrate like w/ saspy, by moving data or variables between SAS and Python, but this would all be SAS code that integrates Python in the workflow. Is that correct? As I try to understand what you're trying to do, I think this is what you're looking for. 

Again, putting this here just trying to really understand what you're trying to do, as much as seeing if the new functionality would solve what you want.

In Viya 4 (I assume you don't have that yet), there is now a PROC PYTHON, which, is SAS code (of course), but integrates Python into SAS, from the SAS language, in a similar, yet reverse, way as SASPy integrates SAS into Python code. You can move data/variables between SAS and the Python process that is created by SAS.

I'm thinking that this would actually be what you are looking for (and correct me if I'm wrong; still trying to understand), in that from EG or any SAS code/program you're putting together, you could do basically the opposite of what I described you doing in SASPy; run you SAS code, data step, whatever, then invoke Proc Python to use simple Python HTTP calls to decode or encode data (which you can move from SAS to the Python process) and then encrypt and move those back to SAS from Python, and then run more data step or whatever to write back to the database. All of this being SAS code you can execute from EG (SAS Workspace server).

Does that make sense, and is that what you are really trying to do with this application? If so, then at least I think I understand now.

Thanks,
Tom

**Comment by pradkum001** [comment]
Hi Tom,
Yes I do understand what you have said.. I am sorry for all the confusion. 
 
My Bad., where ever I said SASPy its just Python script not SASPy. Instead of submitting via SASPy can we call the python script via EG...!

Let me tell you in a different way...

For example take the same query.,

From SAS EG:-

DATA temp;
SET db.patient(KEEP=patient_name patient_address Patient_ID);
RUN;
Where, patient_name and patient_address are the encrypted values and I need to decrypt those two variables from TEMP dataset by calling the script.

After the query above query is done and the temp dataset is created.. 

CALL PYTHON SCRIPT and pass these two variables p_name and p_address on the temp dataset FROM EG.
*cryption happens. 


In this way we are not complicating by involving two clients (EG and SASPy). Its just EG interacting with Python Script (To call authenticate the REST  API)

Hope I am clear now..! The Basic requirement is the end client only uses SAS EG to run a query and they should be able to decrypt the values via EG ONLY. They wont use SASPy nor PYTHON scripts. I initially thought that SASPy now i got to know that they work very differently. 

And I sincerely apologize for using SASPy word again and again instead of python script in my earlier message. I know it confused me too if i read it now. 

Flowchart:
**### _Run SAS QUERY In EG ->>Output ->>Execute Python Script on the output ->>decryption Occurs ->>Final Results_** 


Thank you...

**Comment by tomweber-sas** [comment]
Ah, I think that confirms my recent post. You are just looking to call Python from SAS, so you can use Python HTTP calls instead of using Proc HTTP calls to call the *cryption routines. And yes, that really doesn't have anything to do with saspy. And, Proc Python would solve that quite easily and be exactly what you want. But, you don't have that, so I'm sorry about that. But's it shows a good use case.

The existing Python integration in SAS 9 (as of M6) that you could use to do this would be via either DS2 or Proc FCMP to write functions which you could use to do the decrypt. Actually, the FCMP case may be just what you want for this specific use case. Again, now that I understand, I may have an  answer that works for you.

It's a little complicated, and I don't have the experience with it to write it for you, but you can write Python functions using Proc FCMP, and then use those functions in the data step. It's written to be used for scoring, but it make perfect sense in this case too.  This way they would just be SAS functions you call, but they run Python code.

Take a look at this: https://go.documentation.sas.com/doc/en/pgmsascdc/9.4_3.5/lecompobjref/p18qp136f91aaqn1h54v3b6pkant.htm

And see if this can do what you're looking to do.

Thansk,
Tom

**Comment by pradkum001** [comment]
Tom,
Thank you for all the help and coordination you are doing for me in this process. I really really appreciate your help. 

Coming straight to the point., I am not able to understand what is the need for us to re-write the existing python script into a procedure. Instead, why cant we execute the python script using X command (https://communities.sas.com/t5/SAS-Data-Management/How-to-execute-python-helloworld-py-in-SAS-Enterprise-Guide/td-p/336191) and simultaneously pass the parameters in to it (not sure of how yet but i am sure there should be a way).  I was thinking more of a this approach if that is possible for my requirement. Also., should we pass the python.exe file along to execute within the X command - source - https://communities.sas.com/t5/SAS-Programming/Execute-python-script-within-SAS/td-p/637303.

I am not fully aware if this will work or not but I want to use the python script as is., if its possible instead of incorporating into SAS. 
There are two concerns, firstly I am not sure of how to convert the python script into SAS procedure (FCMP) because there is alot going on in the script., like importing modules (like - requests, json, azure.common.credentials modules ) then going to the URL getting access token, authentication and finally triggering the REST API  and secondly, I am not sure of how long will it take for this process. 

Please let me know if my point makes sense...!

Thank you..

**Comment by tomweber-sas** [comment]
Well, the 'X' command in SAS just let's you get to a shell to execute a command. You can do that, assuming the X command is enabled in the workspace server you're connected to. Again, it's SAS language and that means it's executing in SAS; in the remote workspace server you're connected to. So Python, and the script you execute need to be on that SAS server machine, not you're EG client. machine There is no integration between SAS and whatever command you execute using the X command, so how you move values between these processes is then on you to implement. You can use environment variables; that's easy from both SAS and Python. So, sure you can do it that way if that's what you want to do. I was trying to provide actual integrated support for it that exists in SAS, one way or another.

**Comment by pradkum001** [comment]
Tom, 
I was waiting on  installing modules and in the mean time i was learning and reading more about PROC FCMP (it is very interested and not as complicated as i initially thought). I started trying to convert python code and wrote a basic sample code but to m surprise after running it, I realized that we use an older version of SAS and is not available until M6 Source : (https://documentation.sas.com/doc/en/pgmsascdc/9.4_3.5/basewn/n0ei3jw17d71lin1buwg938z1vit.htm#:~:text=In%20SAS%20Viya%203.5%20and,Python%20function%20using%20PROC%20FCMP).

I checked with the admin team and they say they do not have any plans to update the version.

Now., Am I left with X command Only!!!? Or there any other options ?

Secondly., when I use x command from SAS EG., I was getting an was not getting any error so little i tried to capture the error.

*helloworld.py;
print ("HELLO WORLD!")

*x command;
data test;
x "cd C:\Users\xxxxx\xxxxx";
x "python helloworld.py";
output;
Run;

*capturing the error;
filename oscmd pipe "python  C:\Users\xxxxxx\xxxxx\helloworld.py 2>&1";
data test;
infile oscmd truncover;
input infile $100.;
run;
data _null_;
infile oscmd;
input;
put _infile_;
run;

Error MSG: 'python' is not recognized as an internal or external command,
operable program or batch file.


Python on my system is not in default C:\Programfiles\Python. Since I use VDI it is installed else where...!! 

So., I went to command prompt and tried to submit the same helloworld.py program. I still got the same error. So i did little research and added the python path into the environment variable (source: https://initialcommit.com/blog/python-is-not-recognized-as-an-internal-or-external-command). After adding environment variable path i was able to execute the helloworld.py program via command line.

Now coming back to EG.. I still am getting the same error..!! Does it use any different command prompt ? if yes how to pass the python path and program at the same time?


Thank you,
Pradeep

**Comment by tomweber-sas** [comment]
What machine is EG running on? What machine is SAS (that the EG client is connected to) running on? The X command will execute on the SAS machine. Is python installed on that machine? Is your script accessible from that machine? Try using the full path the the python executable instead of nothing but 'python' and see if the executable is found.

**Comment by tomweber-sas** [comment]
@pradkum001 is there anything else I can do for this issue, or can it be closed? Seems like you don't have the newer versions of SAS that have some of the python integration you need. So you're left to the X command on SAS to run Python scripts directly on the server. I'm not sure there's much else I can do for that scenario. That's a 'user written' use case which doesn't have any integration w/ the SAs system, really.
Thanks,
Tom

**Comment by pradkum001** [comment]
Hi Tom.,  

Yes,  because of the older version I am currently using X command. And sure we can close this for now. 
Appreciate your help in the process Tom.

Thank you so much,
Pradeep

**Comment by tomweber-sas** [comment]
Sure thing, let me know if there's anything else I can do!
Tom
